src/lib/relicScorer.ts

```diff
@@ -126,14 +146,71 @@ export const RelicScorer = {
     return RelicScorer.scoreCharacterWithRelics(character, relics)
   },
 
-  score: (relic, characterId) => {
-    // console.log('score', relic, characterId)
+  scoreRelic: (relic: Relic, id: CharacterId) => {
+    const scoringMetadata = getRelicScoreMeta(id)
+
+    const scoringResult = RelicScorer.score(relic, id)
+    const subScore = parseFloat(scoringResult.score)
+    let mainScore = 0
+    if (Utils.hasMainStat([relic.part])) {
+      if (scoringMetadata.parts[relic.part].includes(relic.main.stat)) {
+        mainScore = 64.8
+      } else {
+        mainScore = scoringMetadata.stats[relic.main.stat] * 64.8
+      }
+    } else {
+      mainScore = 64.8
+    }
+
+    // Predict substat scores
+    const substats = new Set(relic.substats.map((x) => x.stat))
+    const substatScoreEntries = Object.entries(scoringMetadata.stats)
+      .filter((x) => possibleSubstats.has(x[0]))
+      .filter((x) => !substats.has(x[0])) // Exclude already existing substats
+      .sort((a: [string, number], b: [string, number]) => b[1] - a[1])
+
+    const missingSubstats = (4 - substats.size)
+    const missingRolls = Math.ceil(((15 - (5 - relic.grade) * 3) - relic.enhance) / 3) - missingSubstats
+
+    const newSubstats = substatScoreEntries.slice(0, missingSubstats)
+    const finalSubstats = [...substats, ...newSubstats.map((x) => x[0])]
+    const finalSubstatWeights = finalSubstats.map((x) => scoringMetadata.stats[x])
+    const bestOverallSubstatWeight = Math.max(...finalSubstatWeights)
+    const avgWeight = (
+      finalSubstatWeights.reduce((a, b) => a + b, 0)
+      - newSubstats.reduce((a, b) => a + b[1], 0) / 2
+    ) / 4
+
+    let extraRolls = 0
+
+    for (let i = 0; i < missingSubstats; i++) {
+      extraRolls += 1 * newSubstats[i][1]
+    }
+
+    for (let i = 0; i < missingRolls; i++) {
+      extraRolls += bestOverallSubstatWeight
+    }
+
+    const currentWeight = Utils.precisionRound(subScore + mainScore)
+    return {
+      current: currentWeight,
+      best: currentWeight + extraRolls * 6.48,
+      average: currentWeight + extraRolls * 6.48 * avgWeight,
+    }
+  },
 
-    setMainStatFreeRolls()
+  score: (relic, characterId): {
+    score: string
+    rating: string
+    mainStatScore: number
+    part?: number
+    meta?: object
+  } => {
+    // console.log('score', relic, characterId)
 
     if (!relic) {
       return {
-        score: 0,
+        score: '0',
         rating: 'N/A',
```

Comment:
One non blocking note - just want to verify that we actually want this to be a string instead of a number?

Reply:
It is currently treated as a string everywhere and when I tried to treat it as a number things started breaking. So this is just bringing things in alignment.

In terms of whether it *should* be like that...probably not. I'd prefer to send around actual numbers and format when it actually comes to display. But I think I'd suggest making that change after migrating more to typescript so it can pick up issues.